### ISSUE TITLE
Reproducing the Shattering of Gradients

### ISSUE THREAD

**Issue by theSage21** [LSLeary/LL-init#1]
I'm attempting to reproduce the graphs shown in Figure1, namely the increasing tendency towards white noise as number of layers increase.

I've implemented this in PyTorch.

```python
class LabNet(nn.Module):
    def __init__(self, n_layers, hidden_dim=200, fn=None):
        super(LabNet, self).__init__()
        self.__inp = nn.Linear(1, hidden_dim, True)
        self.__hid = [nn.Linear(hidden_dim, hidden_dim, True)
                    for i in range(n_layers)]
        self.__out = nn.Linear(hidden_dim, 1, True)
        
        def rectifier(x):
            return F.relu(x)
        self.__fn = rectifier  if fn is None else fn
        self.__hidden_dim = hidden_dim
        self.__n_layers = n_layers

    def forward(self, x):
        i = self.__fn(self.__inp(x))
        
        for h in self.__hid:
            i = self.__fn(h(i))
        
        o = self.__out(i)
        return o
```

To plot the gradients I'm performing this activity.

```python
def plot_grads(layers):
    input_data = np.linspace(-2, 2, 256)
    input_data = list(map(lambda s: Variable(th.Tensor([s]), requires_grad=True), input_data))
    network = LabNet(layers)
    outputs = []
    for i in input_data:
        network.zero_grad()
        output = network(i)
        output.backward(i)
        outputs.append(i.grad.data.tolist()[0])
    plt.plot(outputs, label=str(layers) + ' layers')
    plt.title(str(activation) if activation is not None else 'Relu')
    plt.legend()
    plt.xlabel('Inputs')
    plt.ylabel('Gradients')
```

With this code I'm able to produce the following two graphs.
![image](https://user-images.githubusercontent.com/7693265/34516052-2bf186c8-f09b-11e7-949d-e62151f824bd.png)

However upon increasing the number of layers, I do not observe the white noise like plot. Am I doing things right? This is the plot I get when I run `plot_grads(24)`.
![image](https://user-images.githubusercontent.com/7693265/34516071-45ccd62e-f09b-11e7-9ece-aedc9ba645f6.png)

**Comment by LSLeary** [comment]
I didn't generate the graphs for figure 1—the code in this repository was only used for section 5, so this isn't really the right place to ask. I'm not sure which of my coauthors generated those plots, but David Balduzzi would know; you should ask him about it. 

All that said, I did generate similar plots for my thesis, so I can give some pointers on observing the effect. I didn't look over your code, but:
  - The points at which the gradient changes are distributed more densely around the origin. Hence taking a smaller window will make the effect more obvious. I used [-1/2,1/2] rather than [-2,2].
  - Similarly, the effect becomes clearer with depth; I used fully connected networks with 99 hidden layers.
  - Wider networks make the output closer to true noise; I used 300 nodes per hidden layer.
  - The appearance of the gradient is sensitive to the variance of the network's weights and biases. I sampled every layer's weights and biases normally with mean 0 and variance 2/N, where N is the number of nodes in the previous layer (1 following the lone input, 300 elsewhere).
